Improves web terminal

maxwang7 opened the pull request:
Issue: #307 

- Allows Ctrl-C
- Captures mouse clicks to move cursor
- Allows Cmd + [number] to switch to a tab (Chrome)
- Make hyperlinks stand out more by using standard CodaLab colors

Concerns: I was having a problem with the focusing on the terminal leading to a re-render of the page that led to some weird problems, so I basically removed focussing. I haven't noticed any problems yet, but it would be best if there was a more well-reasoned approach to getting this to work, or, perhaps less ideal, just testing the command line UI extremely thoroughly to ensure that no functionality broke.

maxwang7 commented:
Ready for review! I tried this again and Ctrl-C is working... Not sure what happened during the demo. Also, I made a more elegant fix to the problem I was facing earlier.

kashizui commented:
Did you apply the customizations made to the old version of jquery-terminal to the new version that you installed? Or is it just the original source?
If the latter, could you try installing it through bower instead? That way there's no need to check in the code into the repo (though it's a valid way to do it in some cases).

kashizui commented on codalab/apps/web/static/js/worksheet/worksheet_action_bar.jsx:
```diff
@@ -48,6 +48,12 @@ var WorksheetActionBar = React.createClass({
       prompt: 'CodaLab> ',
       history: true,
       keydown: function (event, terminal) {
+        if (event.keyCode === 67 && event.ctrlKey) {
```
comment what key 67 refers to

kashizui commented on codalab/apps/web/static/js/worksheet/worksheet_action_bar.jsx:
```diff
@@ -48,6 +48,12 @@ var WorksheetActionBar = React.createClass({
       history: true,
       keydown: function (event, terminal) {
+        if (event.keyCode === 67 && event.ctrlKey) {
+          terminal.exec('', false);
```
What does this line do?
And more generally what is the desired behavior for Ctrl-C on the web terminal? Since it's not a UNIX shell that's sending a SIGINT to the foreground process, probably good to clarify in comments here what the analogous behavior you're trying to emulate here.

percyliang commented:
+1 to installing through bower

On Mon, May 22, 2017 at 10:02 PM, Stephen Koo <notifications@github.com>
wrote:

> *@kashizui* requested changes on this pull request.
>
> Generally looks good, just wanted some clarifications + possibly using
> bower for jquery-terminal.
> ------------------------------
>
> In codalab/apps/web/static/js/worksheet/worksheet_action_bar.jsx
> <https://github.com/codalab/codalab-worksheets/pull/313#discussion_r117897591>
> :
>
> > @@ -48,6 +48,12 @@ var WorksheetActionBar = React.createClass({
>        prompt: 'CodaLab> ',
>        history: true,
>        keydown: function (event, terminal) {
> +        if (event.keyCode === 67 && event.ctrlKey) {
>
> comment what key 67 refers to
> ------------------------------
>
> In codalab/apps/web/static/js/worksheet/worksheet_action_bar.jsx
> <https://github.com/codalab/codalab-worksheets/pull/313#discussion_r117897657>
> :
>
> > @@ -48,6 +48,12 @@ var WorksheetActionBar = React.createClass({
>        prompt: 'CodaLab> ',
>        history: true,
>        keydown: function (event, terminal) {
> +        if (event.keyCode === 67 && event.ctrlKey) {
> +          terminal.exec('', false);
>
> What does this line do?
> And more generally what is the desired behavior for Ctrl-C on the web
> terminal? Since it's not a UNIX shell that's sending a SIGINT to the
> foreground process, probably good to clarify in comments here what the
> analogous behavior you're trying to emulate here.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/codalab/codalab-worksheets/pull/313#pullrequestreview-39640163>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAakuJHnWNbmEGS1B1IPQxfm8M2_mjtpks5r8mhqgaJpZM4Ndo-F>
> .
>

maxwang7 commented:
The behavior I've implemented is creating a new line, leaving the original line (the line on which the user Ctrl-C'd) as is. The way I've implemented it is to:

- The user hits Ctrl-C while on line 1.
- Save the command in line 1.
- Tell the command line to execute an empty command (does nothing).
- Sets the string of line 1 to the original value

It doesn't do any worker-killing, it simply moves the user to the next line.

maxwang7 commented:
Changes: Installs through bower, comment added for 67, tested on local instance.
